### DISCUSSION TITLE
Is it possible to add default annotations for System Managed Components?

### DISCUSSION THREAD

**Discussion by jacky96623** [longhorn/longhorn#6383] [Q&A]
Hi everyone, I am now trying to deploy Longhorn with Helm 3 on an AWS EKS cluster with mixed nodes (EC2 + Fargate). My stragegy is to set up a default Fargate profile (with '*' namespace and no label selector) for generic (stateless) workloads and specifically deploy (stateful) workloads to EC2 nodes with the annotation `eks.amazonaws.com/compute-type: ec2`.

Everything looks okay except that longhorn driver is unstable - and I think the reason is because the pod `discover-proc-kubelet-cmdline` cannot be launched with Fargate due to the privileged mode, etc., settings. I tried to set `defaultSettings.systemManagedComponentsNodeSelector` as `eks.amazonaws.com/capacityType:ON_DEMAND;eks.amazonaws.com/compute-type:ec2` but it cannot solve the issue (maybe the issue arises from scheduler level).

Besides adding label selector to the Fargate profile (which I do not prefer since I want Fargate profile to be default as a fallback), I would like to ask if it is possible to set default annotation similar to the default node selector.

Any help is greatly appreicated!

**Comment by mantissahz** [comment]
I think you also need to add your EC2 nodes into `nodeSelector` for longhorn-driver in https://github.com/longhorn/longhorn/blob/master/chart/values.yaml#L201-L210 such as the tag you mentioned `eks.amazonaws.com/compute-type:ec2`

**Reply by jacky96623** [reply]
Hi @mantissahz , thanks for your prompt reply.

I just tried to also set the same node selectors for the longhorn-driver. Unfortunately, I can only see that the longhorn-driver-deployer pod is scheduled on EC2 instead of the original Fargate but still the `discover-proc-kubelet-cmdline` pod cannot be scheduled under Fargate scheduler.

I recently also found a [blog post](https://ithelp.ithome.com.tw/articles/10294894) talking about some in-depth actions of the Fargate scheduler. It said that the mechanism is achieved by some annotation hooks. I am not familiar with this level of Kubernetes resources but I do believe that allowing extra annotations to system components is somehow beneficial, since it appears to me that the pod is launched from longhorn-driver-deployer but not via K8s manifest.

**Comment by c3y1huang** [comment]
@jacky96623 Could you clarify your use case? Is it either of the following scenarios?

Scenario 1:
1. You want to install Longhorn on both the EC2 nodes and the Fargate nodes.
1. Additionally, you want to somehow configure the Fargate nodes as fallback nodes.

Scenario 2:
1. You want to have Longhorn installed exclusively on the EC2 nodes.

**Reply by jacky96623** [reply]
It is more close to scenario 2 - I want to deploy ingress controller, kubenertes dashboard, etc., with Fargate. Only leaving stateful component like longhorn and MongoDB to be deployed on ec2 nice.

**Reply by c3y1huang** [reply]
Have you tried tainting the Fargate nodes?

**Reply by jacky96623** [reply]
AWS EKS is serverless and each pod is basically launched in somehow dedicated host. The pod is entitled such deployment by their namespace and labels (see selectors in https://docs.aws.amazon.com/eks/latest/userguide/fargate-profile.html#fargate-profile-components).

In short, I think you cannot taint or label it annotate garbage nodes. And as I described earlier, the process is in scheduler level (with admission wehbook etc). There is no use to taint/label/annotate EC2 nodes as well.

**Reply by c3y1huang** [reply]
I see, thank you, @jacky96623 for the clarification.

Longhorn manages component schedulings through the node selector, and this doesn't look like it will work with AWS Fargate when using a wildcard namespace in the profile.

I have created an improvement issue so the team can further investigate this. https://github.com/longhorn/longhorn/issues/6436.
